### PR TITLE
Quoteless metadata

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This small PR adds the ability to declare simple metadata directives without quotation marks (but otherwise leaves quoted metadata as a perfectly fine alternative).

It also renames the tokens the compiler uses for metadata to be more in-line with #828.

### CODE REVIEW

File: cpp/src/Slice/Grammar.y [hunk 1]
```diff
@@ -206,11 +206,11 @@ opt_semicolon
...
 // ----------------------------------------------------------------------
-: ICE_GLOBAL_METADATA_OPEN string_list ICE_GLOBAL_METADATA_CLOSE
+: ICE_FILE_METADATA_OPEN string_list ICE_FILE_METADATA_CLOSE
 {
     $$ = $2;
 }
-| ICE_GLOBAL_METADATA_IGNORE string_list ICE_GLOBAL_METADATA_CLOSE
+| ICE_FILE_METADATA_IGNORE string_list ICE_FILE_METADATA_CLOSE
 {
     unit->error("global metadata must appear before any definitions");
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:215]: Should you fix the message as well?

File: cpp/src/Slice/Grammar.y [hunk 2]
```diff
@@ -220,7 +220,7 @@ global_meta_data
 // ----------------------------------------------------------------------
 meta_data
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:221]: local metadata

File: cpp/src/Slice/Grammar.y [hunk 3]
```diff
@@ -206,11 +206,11 @@ opt_semicolon
 // ----------------------------------------------------------------------
 global_meta_data
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:207]: Inconsistent comment / spelling.

**Comment by InsertCreativityHere** [comment]
Okay, it seems we had different expectations for it's behavior. I had thought that quoteless metadata should be more a drop-in replacement for quoted metadata. So the metadata
```
[[suppress-warning:invalid-metadata, deprecated]]
```
Will be parsed identical to:
```
[["suppress-warning:invalid-metadata", "deprecated"]]
```
Leading and trailing whitespace is trimmed, and brackets, double quotes, and commas are disallowed in unquoted metadata. **But** commas are only disallowed in the sense that if they're present they're treated as delimiters between metadata.

The main difference is that I didn't impose a single-unquoted-metadata per bracket pair like you expected, and instead allowed this. If users need commas or brackets in their metadata in this setup, they'd have to wrap it in quotes.

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:215]: Yes, you're correct about all these, I'll correct the naming for them as well.

**Comment by InsertCreativityHere** [comment]
Reworked the PR from scratch. Now quoteless metadata has the following characteristics:
- Only one metadata directive can be declared in a set of metadata brackets.
- All characters other than " [ ] Are valid within quoteless metadata.
- Leading and trailing whitespace is trimmed from the metadata.
`[     hello< ::there, _you> ]` == `["hello< ::there, _you>"]`

This PR also adds support for concatenating adjacent local metadata (but not file metadata):
`[more]  [ metadata ]` == `["more", "metadata"]`
Apparently having multiple local metadatas wasn't valid Slice!

Finally, it also renames some fields in the compiler to match up with #828, and it updates all our Slice files to use the new quoteless metadata where possible (and fixes some formatting inconsistencies along the way)

**Comment by InsertCreativityHere** [comment]
I mean it's certainly possible to do, but I'm still in the boat that if they need 'special' characters, they should use quoted metadata instead.
Otherwise I think it would be kind of weird, and maybe even misleading, to have this very limited escaping mechanism, but not full escaping support. As in, a user sees you can escape quotes, and might assume it has real escaping support, when it doesn't.

As for allowing quotes in quoteless metadata, again sure it's possible to infer if it's not the first or last character, but I think this again could lead to some flaky and unexpected behavior for users. If they only see quoteless metadata in our examples, and see us using quotes in them safely, they might think that quotes are okay everywhere in metadata, which isn't the case.

So I think we should leave quoteless metadata with the simpler mechanics, and if they need these characters or escaping functionality, they'll have to suffer through typing out two quotation marks, instead of us having this half-baked support for it in quoteless metadata :)

**Comment by bernardnormier** [comment]
With the C# attribute example in the Ice Manual, being able to write:
```[[cs:attribute:assembly: AssemblyDescription("My assembly")]]```
would certainly be nicer than the double-quotes with double-quote escape we currently require:
```[["cs:attribute:assembly: AssemblyDescription(\"My assembly\")"]]```

The syntax is not ambiguous. Maybe I am missing something?

**Comment by InsertCreativityHere** [comment]
I assume the syntax you're talking about is if the first non-whitespace character is a quote, treat it as quoted metadata, otherwise treat it as unquoted?
One question is does this mean we'd be okay with ending quotes?
```
[foo:bar:"whatever"]
```
We would treat this as unquoted then too? Ignoring the fact it's made up metadata.

It's certainly not ambiguous from a technical perspective, but it might catch users off guard who haven't memorized the syntax's oddities, who would expect `["pretty":contrived]` to be valid metadata judging from our examples. And might make it harder to catch errors, since
`[c"cpp:type:MyType"]` would be valid metadata, when it obviously isn't mean to be.

But I'm alot less opposed to allowing double-quotes than I am about the escaping mechanism you were talking about!

**Comment by bernardnormier** [comment]
I would keep it simple:
- if it starts with `["`, is a quoted-metadata
- if it start with `[` and any other character, it's a quoteless metadata. And the compiler just reads everything until it encounters the closing `]`.

Syntactically, `["pretty":contrived]` would not be valid metadata. If it starts with `["`, it must end with `"]`. If it starts with `[` and the next character is not `"`, the character before the closing bracket doesn't matter.

So `[foo:bar:"whatever"]` and `[c"cpp:type:MyType"]` are syntactically fine. The quotes inside the quoteless metadata wouldn't need to be balanced. For example, this is syntactically fine:
`[foo:password:!"?")("]`.

**Comment by InsertCreativityHere** [comment]
Yeah I was saying that `["pretty":contrived]` would indeed be invalid, but that some users might be tricked into expecting it to work, since they might not even know quoted metadata exists by looking through our examples and all.

But alright, I'll 'Okay, implementing it'!

**Comment by InsertCreativityHere** [comment]
Wow, there isn't a single metadata directive that uses escaped double-quotes in it currently?

**Comment by bernardnormier** [comment]
Strangely I am now getting an error for:
```
[[cs:attribute:assembly:AssemblyDescription(\"My assembly\")]]
[[foo:password:!"?")("]]
```

The metadata syntax is correct, but the compiler (`slice2cs`) complains:

```
slice2cs metadata.ice 
/Users/bernard/test/metadata.ice:1: error: Unterminated string literal, skipped the line
    [[cs:attribute:assembly:AssemblyDescription(\"My assembly\")]]
/Users/bernard/test/metadata.ice:2: error: Unterminated string literal, skipped the line
    [[foo:password:!"?")("]]
2 errors in preprocessor.
```

The preprocessor is checking these strings?
For the attribute metadata, I expected to get the \" as-is in the generated code (which would then fail to compile in C#).

**Comment by InsertCreativityHere** [comment]
Yes, I can confirm the same behavior, I didn't check any of these since they would likely be error cases, or at least un-necessary... But you are right that these are still totally valid metadata declarations by our spec.

And the problem is indeed with the preprocessor (MCPP). In the second case it rightly complains the quotes are unbalanced, since you have 3 of them. And in the first case what appears to happen is MCPP only processes escape sequences INSIDE of strings, so it scans the first '\"' as '\"' and starts a string, but then the second '\"' it scans as an escaped quote, and happily continues the string until EOL where it complains...

Looking through our fork, MCPP does do string-aware parsing, and hence requires well-formed strings and quotation matching in source code. I don't see any way to fix this short of replacing MCPP with our own preprocessor, or further hacking our fork of it, but this seems like a waste of time when I intend to replace it anyways... But even in our own preprocessor, I'd planned on having it be string-aware too, and so it would suffer from the same problems unless we also added file-metadata awareness into the preprocessor, but that's starting to get kind of ugly in my book. Or we could drop string-awareness too.

It's all certainly doable, but to your dismay, I'm once again going to say that I think we should leave the 'special' characters for quoted metadata and dissallow them in unquoted metadata :)

**Pull request merged**